judge/views/problem.py

```diff
@@ -201,6 +202,91 @@ def get_context_data(self, **kwargs):
                                           context['description'], 'problem')
         context['meta_description'] = self.object.summary or metadata[0]
         context['og_image'] = self.object.og_image or metadata[1]
+
+        context['vote_perm'] = self.object.vote_permission_for_user(user)
+        if context['vote_perm'].can_vote():
+            try:
+                context['vote'] = ProblemPointsVote.objects.get(voter=user.profile, problem=self.object)
+            except ObjectDoesNotExist:
+                context['vote'] = None
+        else:
+            context['vote'] = None
+
+        return context
+
+
+class ProblemVote(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    http_method_names = ['options', 'post']  # This disables GET requests, even though ProblemMixin.get exists.
+    template_name = 'problem/vote-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if not self.object.vote_permission_for_user(self.request.user).can_vote():
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        try:
+            context['vote'] = ProblemPointsVote.objects.get(voter=self.request.profile, problem=self.object)
+        except ObjectDoesNotExist:
+            context['vote'] = None
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE
+        return context
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.vote_permission_for_user(request.user).can_vote():
+            return HttpResponseForbidden(_('Not allowed to vote on this problem.'), content_type='text/plain')
```

Comment:
I feel like this should also be a JSON response.

Reply:
already resolved, this is currently `return JsonResponse`